server/src/main/java/org/eclipse/hono/application/HonoApplication.java

```diff
@@ -0,0 +1,248 @@
+
+import org.eclipse.hono.authentication.AuthenticationService;
+import org.eclipse.hono.authorization.AuthorizationService;
+import org.eclipse.hono.config.HonoConfigProperties;
+import org.eclipse.hono.registration.RegistrationService;
+import org.eclipse.hono.server.HonoServer;
+import org.eclipse.hono.server.HonoServerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+
+/**
+ * The Hono server main application class.
+ * <p>
+ * This class configures and wires up Hono's components (Verticles).
+ * By default there will be as many instances of each verticle created as there are CPU cores
+ * available. The {@code hono.maxinstances} config property can be used to set the maximum number
+ * of instances to create. This may be useful for executing tests etc.
+ * </p>
+ */
+@Component
+public class HonoApplication {
+
+    private static final Logger LOG = LoggerFactory.getLogger(HonoApplication.class);
+
+    private HonoConfigProperties honoConfig;
+    private Vertx vertx;
+    private RegistrationService registrationService;
+    private AuthenticationService authenticationService;
+    private AuthorizationService authorizationService;
+    private HonoServerFactory serverFactory;
+
+    /**
+     * @param honoConfig the honoConfig to set
+     */
+    @Autowired
+    public void setHonoConfig(final HonoConfigProperties honoConfig) {
+        this.honoConfig = honoConfig;
+    }
+
+    /**
+     * @param vertx the vertx to set
+     */
+    @Autowired
+    public void setVertx(final Vertx vertx) {
+        this.vertx = vertx;
+    }
+
+    /**
+     * @param registrationService the registrationService to set
+     */
+    @Autowired
+    public void setRegistrationService(final RegistrationService registrationService) {
+        this.registrationService = registrationService;
+    }
+
+    /**
+     * @param authenticationService the authenticationService to set
+     */
+    @Autowired
+    public void setAuthenticationService(final AuthenticationService authenticationService) {
+        this.authenticationService = authenticationService;
+    }
+
+    /**
+     * @param authorizationService the authorizationService to set
+     */
+    @Autowired
+    public void setAuthorizationService(final AuthorizationService authorizationService) {
+        this.authorizationService = authorizationService;
+    }
+
+    /**
+     * @param serverFactory the serverFactory to set
+     */
+    @Autowired
+    public void setServerFactory(final HonoServerFactory serverFactory) {
+        this.serverFactory = serverFactory;
+    }
+
+    /**
+     * Deploys all verticles the Hono server consists of.
+     */
+    @PostConstruct
+    public void registerVerticles() {
```

Comment:
should be made final, you may also want to provide a separate, empty, protected method e.g. like `postRegisterVeticles()` which gets invoked from here and which subclasses can override in order to customize the application?
Alternatively, the `@PostConstruct` annotation should be removed so that the subclass is in control over the start up process ...